Collapse MutableByteSequence into ArrayByteSequence

meatballspaghetti opened the pull request:
This PR is meant to delete the `MutableByteSequence` class which was used by 4 files: `RFile`, `LocalityGroupUtil`, `RelativeKey`, and `RelativeKeyTest`. Upon deleting the class, its usages were replaced with the parent class `ArrayByteSequence`, which maintains the functionality from previous `MutableByteSequence` references.

The following files have been modified:
- `core/src/main/java/org/apache/accumulo/core/util/MutableByteSequence.java`: Deleted.
- `core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java`:  Migrate byte sequence modification methods from `MutableByteSequence`. <s>Create `forMutableByteSequence()` method, a static named constructor to replace the functionality of previous `MutableByteSequence` objects. Functionality included:
  - Creates a `ArrayByteSequence` object
  - Disallows `ArrayByteSequence` objects from being modified, effectively maintaining their immutability, unless they were created by `forMutableByteSequence()` constructor </s>
- `core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java`,
  `core/src/main/java/org/apache/accumulo/core/file/rfile/RelativeKey.java`,
  `core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java`, and
  `core/src/test/java/org/apache/accumulo/core/file/rfile/RelativeKeyTest.java`
  have each had their previous references to `MutableByteSequence` replaced with `ArrayByteSequence`.

- UPDATE: New `forMutableByteSequence()` constructors have been removed. Previous references to `MutableByteSequence` have been replaced with just `ArrayByteSequence`. `ArrayByteSequence` was never completely immutable in the first place, so functionality is maintained.

Fixes #4738 "Look at collapsing MutableByteSequence into ArrayByteSequence"

dlmarion commented:
@meatballspaghetti - I think you should rebase these changes onto `main`. Reason being that `ArrayByteSequence.reset` was added in 3.1 and overlaps the functionality of `setArray` and `setLength`. I think when you rebase onto `main`, your additions may not be needed at all. I think MutableByteSequence can probably be deleted and the other classes fixed to use the existing ArrayByteSequence methods. I don't think that the `isMutable` member variable needs to be added as it gives a false sense of immutability. Some of the constructors in ArrayByteSequence retain a reference to the underlying byte[] instead of copying it. If that underlying byte[] changes, then the contents of this ArrayByteSequence change.

meatballspaghetti commented:
> @meatballspaghetti - I think you should rebase these changes onto `main`. Reason being that `ArrayByteSequence.reset` will be added in 3.1 (#4735) and overlaps the functionality of `setArray` and `setLength`. I think when you rebase onto `main`, your additions may not be needed at all. I think MutableByteSequence can probably be deleted and the other classes fixed to use the existing ArrayByteSequence methods. I don't think that the `isMutable` member variable needs to be added as it gives a false sense of immutability. Some of the constructors in ArrayByteSequence retain a reference to the underlying byte[] instead of copying it. If that underlying byte[] changes, then the contents of this ArrayByteSequence change.

Okay, makes sense. I'll switch to `main` and move forward with just the deletion of `MutableByteSequence`, without the additions from this branch.

meatballspaghetti commented:
Updated to change base branch to `main` and remove previous constructor additions. All previous references to `MutableByteSequence` now replaced with `ArrayByteSequence`.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java:
```diff
@@ -180,4 +180,14 @@ public byte[] toArray() {
     return new String(data, offset, length, UTF_8);
   }
+
+  public void setArray(byte[] data, int offset, int len) {
+    this.data = data;
+    this.offset = offset;
+    this.length = len;
+  }
+
+  public void setLength(int len) {
+    this.length = len;
+  }
```
In the constructor, there are bounds checks to make sure that the length and offset values make sense. Those same checks should appear here or else there will be exceptions later.

Similarly, the setLength parameter should be checked to make sure it's a valid length.

meatballspaghetti replied:
Done, updated.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:
```diff
@@ -993,7 +992,7 @@ private void _seek(Range range) throws IOException {
           // and speed up others.
 
-          MutableByteSequence valbs = new MutableByteSequence(new byte[64], 0, 0);
+          ArrayByteSequence valbs = new ArrayByteSequence(new byte[64], 0, 0);
```
For the lines that changed, you could use `var` to make it slightly cleaner, since the type is redundant in these declarations.

```suggestion
          var valbs = new ArrayByteSequence(new byte[64], 0, 0);
```

meatballspaghetti replied:
Done, updated.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java:
```diff
@@ -180,4 +180,21 @@ public byte[] toArray() {
     return new String(data, offset, length, UTF_8);
   }
+
+  public void setArray(byte[] data, int offset, int length) {
```
This needs javadoc with a `@since 3.1.0` tag.

meatballspaghetti replied:
Done, updated.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java:
```diff
@@ -180,4 +180,21 @@ public byte[] toArray() {
+    this.length = length;
+  }
+
+  public void setLength(int length) {
+    if (length < 0) {
+      throw new IllegalArgumentException(" Bad length = " + length);
+    }
+    this.length = length;
+  }
```
Wondering if we can drop this method, not sure.   Its only called in two places, wonder if it would be cleaner to refactor those two places to call `setArray` instead of this setLength method.  

If this method is kept then its validation will need to be improved to ensure that length makes sense for the current offset and data.length.   Also need to add javadoc that states it sets the length if the length is less than the length of the current array minus the current offset.  Also the javadoc would need to include a `@since 3.1.0` tag.

meatballspaghetti replied:
Agreed that it can be dropped. Updated.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java:
```diff
@@ -180,4 +180,21 @@ public byte[] toArray() {
     return new String(data, offset, length, UTF_8);
   }
+
+  public void setArray(byte[] data, int offset, int length) {
```
Feel free to ignore this comment as its personal preference.  I like the method name `reset` used in #4735, so could use that instead of `setArray`. This is a new public API method is being introduced, so can call it whatever we like.  Do not need to use the same method name as MutableByteSequence

meatballspaghetti replied:
Sounds good to me, I renamed `setArray` to `reset`.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RelativeKey.java:
```diff
@@ -432,24 +432,24 @@ public static SkippR fastSkip(DataInput in, Key seekKey, MutableByteSequence val
       throws IOException {
     if (mbseqDestination.getBackingArray().length < len) {
-      mbseqDestination.setArray(new byte[UnsynchronizedBuffer.nextArraySize(len)], 0, 0);
+      mbseqDestination.reset(new byte[UnsynchronizedBuffer.nextArraySize(len)], 0, 0);
     }
 
     in.readFully(mbseqDestination.getBackingArray(), 0, len);
-    mbseqDestination.setLength(len);
+    mbseqDestination.reset(mbseqDestination.getBackingArray(), mbseqDestination.offset(), len);
```
I think this could be redone as:

```
    if (mbseqDestination.getBackingArray().length < len) {
      byte[] buf = new byte[UnsynchronizedBuffer.nextArraySize(len);
      in.readFully(buf, 0, len);
      mbseqDestination.reset(buf, 0, len);
    } else {
      in.readFully(mbseqDestination.getBackingArray(), 0, len);
      mbseqDestination.reset(mbseqDestination.getBackingArray(), mbseqDestination.offset(), len);
    }
```

dlmarion replied:
> I think this could be redone as:
> 
> ```
>     if (mbseqDestination.getBackingArray().length < len) {
>       byte[] buf = new byte[UnsynchronizedBuffer.nextArraySize(len);
>       in.readFully(buf, 0, len);
>       mbseqDestination.reset(buf, 0, len);
>     } else {
>       in.readFully(mbseqDestination.getBackingArray(), 0, len);
>       mbseqDestination.reset(mbseqDestination.getBackingArray(), mbseqDestination.offset(), len);
>     }
> ```

Actually, it can be further simplified:

```
    byte[] buf = dest.getBackingArray();
    if (dest.getBackingArray().length < len) {
      buf = new byte[UnsynchronizedBuffer.nextArraySize(len)];
    }
    in.readFully(buf, 0, len);
    mbseqDestination.reset(buf, 0, len);
```

keith-turner replied:
Could simplify a bit further.

```java
    byte[] buf = dest.getBackingArray();
    if (buf.length< len) {
      buf = new byte[UnsynchronizedBuffer.nextArraySize(len)];
    }
    in.readFully(buf, 0, len);
    mbseqDestination.reset(buf, 0, len);
```

meatballspaghetti replied:
Thanks much, updated.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RelativeKey.java:
```diff
@@ -467,13 +467,13 @@ private static byte[] readPrefix(DataInput in, ByteSequence prefixSource) throws
     int remainingLen = WritableUtils.readVInt(in);
     int len = prefixLen + remainingLen;
     if (dest.getBackingArray().length < len) {
-      dest.setArray(new byte[UnsynchronizedBuffer.nextArraySize(len)], 0, 0);
+      dest.reset(new byte[UnsynchronizedBuffer.nextArraySize(len)], 0, 0);
     }
```
You could probably do the following here:

```
    byte[] buf = dest.getBackingArray();
    if (dest.getBackingArray().length < len) {
      buf = new byte[UnsynchronizedBuffer.nextArraySize(len)];
    } 
```

Then, everywhere below where `dest.getBackingArray` is used in the calls to `System.arraycopy`, replace them with `buf`. Then finally call `ArrayByteSequence.reset` using `buf`.

meatballspaghetti replied:
Updated.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java:
```diff
@@ -180,4 +180,19 @@ public byte[] toArray() {
+   *
+   * @since 3.1.0
+   */
+  public void reset(byte[] data, int offset, int length) {
```
#4735 has been merged into `main`. The reset method now exists, but is missing the bounds checking that you have added here.

dlmarion commented:
@meatballspaghetti - looks like there is a conflict in the reset method. You will need to merge `main` into your feature branch, resolve conflicts, and re-push.

meatballspaghetti commented:
> @meatballspaghetti - looks like there is a conflict in the reset method. You will need to merge `main` into your feature branch, resolve conflicts, and re-push.

Done.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RelativeKey.java:
```diff
@@ -467,24 +467,24 @@ private static byte[] readPrefix(DataInput in, ByteSequence prefixSource) throws
     int remainingLen = WritableUtils.readVInt(in);
     int len = prefixLen + remainingLen;
+    byte[] buf = dest.getBackingArray();
     if (dest.getBackingArray().length < len) {
```
Could change this to `buf.length` like the code above.

meatballspaghetti replied:
Done.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RelativeKey.java:
```diff
@@ -467,24 +467,24 @@ private static byte[] readPrefix(DataInput in, ByteSequence prefixSource) throws
-    in.readFully(dest.getBackingArray(), prefixLen, remainingLen);
-    dest.setLength(len);
+    in.readFully(buf, prefixLen, remainingLen);
+    dest.reset(buf, dest.offset(), len);
```
is `dest.offset()` correct here? I'm curious why it's not zero.

meatballspaghetti replied:
Agreed that it should be 0, updated.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java:
```diff
@@ -160,6 +160,10 @@ public int offset() {
    * @since 3.1.0
    */
   public void reset(byte[] data, int offset, int length) {
+    if (offset < 0 || offset > data.length || length < 0 || (offset + length) > data.length) {
+      throw new IllegalArgumentException(" Bad offset and/or length data.length = " + data.length
+          + " offset = " + offset + " length = " + length);
+    }
```
@keith-turner identified [on another issue](https://github.com/apache/accumulo/pull/4524#discussion_r1692342340) that Java's built-in [Objects](https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Objects.html) has some methods to check bounds as well.

The exceptions thrown for each of the 3 (what we have, Objects, Preconditions) are all slightly different, but are all RuntimeExceptions, so fundamentally they all should check things the same way... it would only matter if somebody had a specific catch block checking for which RTE it was.

Of the options, I'm inclined to choose [the built-in Objects version][1].

[1]: https://docs.oracle.com/en%2Fjava%2Fjavase%2F11%2Fdocs%2Fapi%2F%2F/java.base/java/util/Objects.html#checkFromIndexSize(int,int,int)

meatballspaghetti replied:
Updated to use the Objects methods. They throw IndexOutOfBoundsException instead, which I updated in the unit tests in ArrayByteSequenceTest as well.

meatballspaghetti commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:
```diff
@@ -1055,7 +1054,7 @@ private void _seek(Range range) throws IOException {
           }
 
-          MutableByteSequence valbs = new MutableByteSequence(new byte[64], 0, 0);
+          ArrayByteSequence valbs = new ArrayByteSequence(new byte[64], 0, 0);
```
Done.

meatballspaghetti commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:
```diff
@@ -1072,7 +1071,7 @@ private void _seek(Range range) throws IOException {
 
                 val.readFields(currBlock);
-                valbs = new MutableByteSequence(val.get(), 0, val.getSize());
+                valbs = new ArrayByteSequence(val.get(), 0, val.getSize());
```
Done.

ctubbsii reviewed (approved):
Looks really good to me. The only other thing I'd suggest is maybe do a static import on Objects

meatballspaghetti commented on core/src/main/java/org/apache/accumulo/core/data/ArrayByteSequence.java:
```diff
@@ -168,10 +161,7 @@ public void reset(byte[] data, int offset, int length) {
   @Override
   public ByteSequence subSequence(int start, int end) {
 
-    if (start > end || start < 0 || end > length) {
-      throw new IllegalArgumentException("Bad start and/end start = " + start + " end=" + end
-          + " offset=" + offset + " length=" + length);
-    }
```
Is `length` correct here, or should it changed to `data.length`? Wasn't completely sure.

ctubbsii replied:
You commented on the wrong line in GitHub, so it's not showing me the line you're asking about. If you want to comment on a range of lines, you can click on the first line and drag down until the last line, and it will show the whole chunk. To your question, I think you did it correctly. `length` refers to the length of the limited "view" of the backing array, and the start and end of the subsequence are coming from that "view" of the backing array, not the full backing array, and have to fit in that view's length.

meatballspaghetti commented:
> Looks really good to me. The only other thing I'd suggest is maybe do a static import on Objects

I would have done static import but I use 3 different bounds checking methods from Objects in this commit. Should I static import all 3 of them anyways, or leave as is?

ctubbsii commented:
> > Looks really good to me. The only other thing I'd suggest is maybe do a static import on Objects
> 
> I would have done static import but I use 3 different bounds checking methods from Objects in this commit. Should I static import all 3 of them anyways, or leave as is?

I think it's pretty much the same either way. It's fine to leave it as is.

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:
```diff
@@ -1055,7 +1054,7 @@ private void _seek(Range range) throws IOException {
           }
 
-          MutableByteSequence valbs = new MutableByteSequence(new byte[64], 0, 0);
+          final ArrayByteSequence valbs = new ArrayByteSequence(new byte[64], 0, 0);
```
```suggestion
          final var valbs = new ArrayByteSequence(new byte[64], 0, 0);
```

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java:
```diff
@@ -993,7 +992,7 @@ private void _seek(Range range) throws IOException {
           // and speed up others.
 
-          MutableByteSequence valbs = new MutableByteSequence(new byte[64], 0, 0);
+          var valbs = new ArrayByteSequence(new byte[64], 0, 0);
```
```suggestion
          final var valbs = new ArrayByteSequence(new byte[64], 0, 0);
```

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/util/LocalityGroupUtil.java:
```diff
@@ -306,7 +306,7 @@ public Partitioner(PreAllocatedArray<Map<ByteSequence,MutableLong>> groups) {
         PreAllocatedArray<List<Mutation>> partitionedMutations) {
 
-      MutableByteSequence mbs = new MutableByteSequence(new byte[0], 0, 0);
+      ArrayByteSequence mbs = new ArrayByteSequence(new byte[0], 0, 0);
```
```suggestion
      final var mbs = new ArrayByteSequence(new byte[0], 0, 0);
```
